File: windows/tests/TraceCommons.Interop.Tests/ShellWordingTests.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using Xunit;

namespace TraceCommons.Interop.Tests;

/// <summary>
/// Wording authored in this shell, over the whole shell.
///
/// <para>
/// Two guards already hold single surfaces to the strict rule -- every string
/// literal in <c>RoutingTools.cs</c>, and in the four witness files, must be a
/// wire value, because the sentence beside it crosses the ABI already
/// finished. Those stay exactly as they are. What they could not see is a
/// sentence hand-written in any of the other hundred-odd files: wording
/// authored in one shell survives a rename in the other two, and nothing here
/// noticed.
/// </para>
/// <para>
/// This reads every C# and XAML source of both shell projects and counts the
/// literals that read as a sentence a contributor would be shown. It is a
/// ratchet, not a clean bill of health: the Windows shell today authors most
/// of its own wording -- the <c>*Copy</c> classes in the interop assembly were
/// transcribed from the shared design, and the XAML views carry their own
/// <c>Text</c> and <c>ToolTipService.ToolTip</c> strings. Moving all of that
/// behind the ABI is a project of its own. Every file that does so is recorded
/// below with the exact number of sentences it holds today, so nothing new can
/// be added to it and nothing new can start doing it.
/// </para>
/// </summary>
public class ShellWordingTests
{
    /// <summary>
    /// Files that author wording today, and exactly how much.
    ///
    /// <para>
    /// TODO(shell-copy): every entry here is a file whose wording should be
    /// composed in the Rust contributor crate and read across the C ABI, the
    /// way <c>routing_copy.rs</c> and <c>witness_copy.rs</c> already are.
    /// Until then the number is a CEILING AND A FLOOR both: adding a sentence
    /// fails, and removing one fails too, so that the entry has to be lowered
    /// deliberately as copy moves out. Never raise a number. A new file must
    /// never be added here.
    /// </para>
    /// </summary>
    private static readonly IReadOnlyDictionary<string, int> WordingBaseline =
        new SortedDictionary<string, int>(StringComparer.Ordinal)
        {
            // The interop assembly's copy classes. Each is a transcription of
            // the shared design's wording, kept in the interop assembly rather
            // than in a view model so a machine that cannot build WinUI can
            // still test it -- which is precisely why the same sentences are
            // written a second time in the GTK and macOS shells today.
            { "TraceCommons.Interop/ArmingOffer.cs", 4 },
            { "TraceCommons.Interop/CorrectionCopy.cs", 4 },
            { "TraceCommons.Interop/HealthCopy.cs", 22 },
            { "TraceCommons.Interop/HistoryCopy.cs", 30 },
            { "TraceCommons.Interop/OriginalSearchOutcome.cs", 4 },
            { "TraceCommons.Interop/PreviewCardOutcome.cs", 1 },
            { "TraceCommons.Interop/ProjectIgnoreCopy.cs", 8 },
            { "TraceCommons.Interop/PublicProfileCopy.cs", 44 },
            { "TraceCommons.Interop/RedactionLabels.cs", 3 },
            { "TraceCommons.Interop/RedactionSummary.cs", 7 },
            { "TraceCommons.Interop/ScrubDetectorCopy.cs", 6 },
            { "TraceCommons.Interop/ScrubbingCaveatCopy.cs", 3 },
            { "TraceCommons.Interop/SessionRootsCopy.cs", 14 },
            { "TraceCommons.Interop/SubagentCopy.cs", 2 },
            { "TraceCommons.Interop/SubmitToast.cs", 10 },
            { "TraceCommons.Interop/TrayModel.cs", 11 },
            { "TraceCommons.Interop/UnresolvedBucketCopy.cs", 3 },
            { "TraceCommons.Interop/UpdateProtocol.cs", 10 },
            { "TraceCommons.Interop/VerdictCopy.cs", 4 },
            { "TraceCommons.Interop/WatchCopy.cs", 4 },
            { "TraceCommons.Interop/WeekBandCopy.cs", 1 },
            { "TraceCommons.Interop/WithdrawCopy.cs", 34 },

            // The read gate. Its four sentences are the safety claim a
            // contributor reads before approving, and they are the highest
            // priority of everything on this list: a claim about what leaves
            // the machine must not be written three times in three shells.
            { "TraceCommons.Interop/ReadGate.cs", 4 },

            // View models that compose a sentence rather than reading one.
            // ContributorSettingsViewModel is the file the settings-screen
            // guard already watches for the witness row specifically; the rest
            // of its wording is unmoved.
            { "TraceCommons.App/ViewModels/ContributorSettingsViewModel.cs", 20 },
            { "TraceCommons.App/ViewModels/HistoryViewModel.cs", 4 },
            { "TraceCommons.App/ViewModels/MainViewModel.cs", 20 },
            { "TraceCommons.App/ViewModels/OnboardingViewModel.cs", 6 },
            { "TraceCommons.App/ViewModels/PreviewSheetViewModel.cs", 10 },
            { "TraceCommons.App/ViewModels/QueueGroupViewModel.cs", 1 },
            { "TraceCommons.App/ViewModels/SessionRootsViewModel.cs", 2 },

            // Window and control code-behind: dialog bodies and one fallback
            // label, written at the call site.
            { "TraceCommons.App/MainWindow.xaml.cs", 3 },
            { "TraceCommons.App/StartupRegistration.cs", 4 },
            { "TraceCommons.App/TrayIcon.cs", 3 },

            // XAML views. Literal Text=, Header=, PlaceholderText= and
            // ToolTipService.ToolTip= content -- the accessibility labels
            // included, which are as much a rename risk as anything drawn.
            { "TraceCommons.App/Controls/HistoryView.xaml", 8 },
            { "TraceCommons.App/Controls/PreviewSheet.xaml", 20 },
            { "TraceCommons.App/Controls/SettingsView.xaml", 16 },
            { "TraceCommons.App/MainWindow.xaml", 19 },
            { "TraceCommons.App/OnboardingWindow.xaml", 18 },
            { "TraceCommons.App/SessionRootsWindow.xaml", 3 },
        };

    /// <summary>
    /// The surfaces whose wording already comes from Rust. Nothing may ever
    /// buy them an allowance here: five of the six are held to the strict
    /// every-literal rule by <c>NoWordingIsAuthoredInThisShell</c> and
    /// <c>NoWordingIsAuthoredInTheWitnessSurface</c>, and a baseline entry
    /// would be a quieter way of undoing that. The sixth,
    /// <c>RoutingSurface.cs</c>, is the routing surface's other half: no
    /// strict guard names it, so zero here is the only thing holding it.
    /// </summary>
    private static readonly string[] RustOwnedSurfaces =
    {
        "TraceCommons.Interop/RoutingTools.cs",
        "TraceCommons.Interop/RoutingSurface.cs",
        "TraceCommons.Interop/WitnessTools.cs",
        "TraceCommons.Interop/WitnessSurface.cs",
        "TraceCommons.Interop/NearAccountConnection.cs",
        "TraceCommons.Interop/AdmissionPreparation.cs",
    };

    /// <summary>
    /// Words a sentence has and an identifier, a wire key, a resource key or a
    /// format pattern does not.
    ///
    /// <para>
    /// This is what separates authored prose from the rest, and it is
    /// deliberately a function-word test rather than a "has a space" test:
    /// <c>"Segoe UI Variable Display"</c>, <c>"MMMM d, yyyy"</c> and
    /// <c>"0,0,0,12"</c> are not wording, and <c>"Watch this folder"</c>,
    /// <c>"Still being scored"</c> and <c>"Back to the folder list"</c> are.
    /// </para>
    /// </summary>
    private static readonly HashSet<string> FunctionWords = new(
        @"a an and are as at be been being but by can cannot could did do does for from
          had has have how if in into is isn't it it's its just may never no not nothing of off on once only or
          so some still such than that the their them then there they this those to until up was we were what
          when where which while who will with would you your yours yet anything something everything already
          always about after again all any because before both each else ever every here more most much
          must need needs same see should since take takes tell these too under use used using very"
            .Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries),
        StringComparer.Ordinal);

    private static readonly Regex StringLiteral = new("\"([^\"\\\\]|\\\\.)*\"", RegexOptions.Compiled);

    private static readonly Regex CharLiteral = new("'(\\\\.|[^'\\\\])'", RegexOptions.Compiled);

    private static readonly Regex XamlComment = new("<!--.*?-->", RegexOptions.Compiled | RegexOptions.Singleline);

    private static readonly Regex XamlAttributeValue = new("=\"([^\"]*)\"", RegexOptions.Compiled);

    private static readonly Regex XamlTextNode = new(">([^<>]+)<", RegexOptions.Compiled);

    /// <summary>
    /// No file in this shell authors more wording than it did when this guard
    /// was written, and no file starts authoring wording that did not.
    /// </summary>
    [Fact]
    public void NoWordingIsAuthoredInThisShellBeyondTheRecordedBaseline()
    {
        var scanned = ScanShellSources();

        // A copy that silently stopped happening would turn this test into a
        // pass over nothing, which is the failure mode the single-file guards
        // already protect against by name.
        Assert.True(
            scanned.Count >= 100,
            $"only {scanned.Count} shell sources were copied to shell-source/; "
            + "the whole tree of both projects is expected. See the csproj.");

        var failures = new List<string>();

        foreach (var (relativePath, wording) in scanned.OrderBy(pair => pair.Key, StringComparer.Ordinal))
        {
            int allowed = WordingBaseline.TryGetValue(relativePath, out int budget) ? budget : 0;
            if (wording.Count == allowed)
            {
                continue;
            }

            if (wording.Count > allowed)
            {
                failures.Add(
                    $"{relativePath}: {wording.Count} authored sentences, baseline allows {allowed}. "
                    + $"First one over the line: \"{wording[allowed]}\"");
            }
            else
            {
                failures.Add(
                    $"{relativePath}: {wording.Count} authored sentences, baseline still allows {allowed}. "
                    + "Wording moved out -- lower the entry (or delete it at zero).");
            }
        }

        foreach (string recorded in WordingBaseline.Keys)
        {
            if (!scanned.ContainsKey(recorded))
            {
                failures.Add($"{recorded}: recorded in the baseline but no longer in the shell. Delete the entry.");
            }
        }

        Assert.True(
            failures.Count == 0,
            "Wording on this shell's surfaces comes from the Rust contributor crate across the ABI.\n"
            + "A sentence written here is one the other two shells will not get, and one a rename in\n"
            + "the Rust will not reach.\n\n"
            + string.Join("\n", failures));
    }

    /// <summary>
    /// The surfaces the Rust already owns hold no wording at all, and hold no
    /// baseline entry either.
    /// </summary>
    [Fact]
    public void TheRustOwnedSurfacesAreNotGivenAWordingAllowance()
    {
        var scanned = ScanShellSources();

        foreach (string surface in RustOwnedSurfaces)
        {
            Assert.False(
                WordingBaseline.ContainsKey(surface),
                $"{surface} has a wording baseline entry. Its wording comes from Rust and the strict "
                + "per-literal guard says so; an allowance here would quietly undo that.");

            Assert.True(
                scanned.ContainsKey(surface),
                $"{surface} was not among the copied shell sources; the guard would pass over nothing.");

            Assert.Empty(scanned[surface]);
        }
    }

    /// <summary>
    /// Reads every copied shell source and returns the wording each one
    /// authors, keyed by its path relative to <c>windows/src</c>.
    /// </summary>
    private static Dictionary<string, List<string>> ScanShellSources()
    {
        string root = Path.Combine(AppContext.BaseDirectory, "shell-source");
        Assert.True(Directory.Exists(root), $"the shell sources were not copied to {root}");

        var scanned = new Dictionary<string, List<string>>(StringComparer.Ordinal);
        foreach (string file in Directory.EnumerateFiles(root, "*.txt", SearchOption.AllDirectories))
        {
            string relative = Path.GetRelativePath(root, file).Replace('\\', '/');
            relative = relative[..^".txt".Length];
            scanned[relative] = relative.EndsWith(".xaml", StringComparison.Ordinal)
                ? AuthoredWordingInXaml(File.ReadAllText(file))
                : AuthoredWordingInCSharp(File.ReadAllText(file));
        }

        return scanned;
    }

    /// <summary>
    /// The sentences a C# source authors.
    /// </summary>
    /// <remarks>
    /// Comment lines go first, for the reason the routing guard gives: prose
    /// about the wire may quote it, and nothing in a comment is rendered.
    /// Char literals go next so that a <c>'"'</c> cannot unbalance the literal
    /// scan. What is left is filtered by statement: a message handed to
    /// <c>throw</c> or to <c>Debug.WriteLine</c> is read by whoever is
    /// debugging this and by nobody else, and holding those to the shared copy
    /// would be a refinement of nothing.
    /// </remarks>
    private static List<string> AuthoredWordingInCSharp(string source)
    {
        string uncommented = string.Join(
            "\n",
            source.Split('\n').Where(line => !line.TrimStart().StartsWith("//", StringComparison.Ordinal)));
        uncommented = CharLiteral.Replace(uncommented, "''");

        var wording = new List<string>();
        foreach (Match match in StringLiteral.Matches(uncommented))
        {
            if (IsDeveloperFacing(uncommented, match.Index))
            {
                continue;
            }

            string literal = match.Value[1..^1];
            if (ReadsAsASentence(literal))
            {
                wording.Add(literal);
            }
        }

        return wording;
    }

    /// <summary>
    /// The sentences a XAML source authors: attribute values and text nodes
    /// both, since <c>&lt;TextBlock&gt;Nothing yet&lt;/TextBlock&gt;</c> draws
    /// exactly what <c>Text="Nothing yet"</c> does.
    /// </summary>
    private static List<string> AuthoredWordingInXaml(string source)
    {
        string uncommented = XamlComment.Replace(source, string.Empty);

        var wording = new List<string>();
        foreach (Match match in XamlAttributeValue.Matches(uncommented))
        {
            if (ReadsAsASentence(match.Groups[1].Value))
            {
                wording.Add(match.Groups[1].Value);
            }
        }

        foreach (Match match in XamlTextNode.Matches(uncommented))
        {
            if (ReadsAsASentence(match.Groups[1].Value))
            {
                wording.Add(match.Groups[1].Value);
            }
        }

        return wording;
    }

    /// <summary>
    /// True where the literal at <paramref name="index"/> is an exception
    /// message, a debug line or a <c>nameof</c> neighbour rather than
    /// something a contributor is shown.
    /// </summary>
    private static bool IsDeveloperFacing(string source, int index)
    {
        int boundary = source.LastIndexOfAny(new[] { ';', '{', '}' }, Math.Max(index - 1, 0));
        string statement = source[(boundary + 1)..index];
        return statement.Contains("throw", StringComparison.Ordinal)
            || statement.Contains("Debug.WriteLine", StringComparison.Ordinal)
            || statement.Contains("Debug.Assert", StringComparison.Ordinal)
            || statement.Contains("nameof", StringComparison.Ordinal);
    }

    /// <summary>
    /// True where the literal reads as a sentence somebody wrote for a
    /// contributor to read.
    /// </summary>
    private static bool ReadsAsASentence(string literal)
    {
        string text = literal.Replace('\n', ' ').Replace('\r', ' ');
        if (!text.Contains(' ', StringComparison.Ordinal))
        {
            return false;
        }

        // A binding, a markup extension or a resource reference. Whatever it
        // resolves to is somebody else's literal and is counted there.
        if (text.TrimStart().StartsWith("{", StringComparison.Ordinal))
        {
            return false;
        }

        var words = text
            .Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries)
            .Select(LettersOnly)
            .Where(word => word.Length > 0)
            .ToList();

        return words.Count >= 2 && words.Any(FunctionWords.Contains);
    }

    private static string LettersOnly(string token)
    {
        var builder = new StringBuilder(token.Length);
        foreach (char c in token)
        {
            if ((c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z') || c == '\'')
            {
                builder.Append(char.ToLowerInvariant(c));
            }
        }

        return builder.ToString();
    }
}

File: macos/Tests/TCShellCoreTests/ShellWordingTests.swift
import XCTest

/// Wording authored in this shell, over the whole shell.
///
/// The mirror of `ShellWordingTests.cs` on Windows, for the same reason and
/// with the same shape: a sentence hand-written in one shell survives a
/// rename in the other two, and until this file existed nothing here
/// noticed. Every Swift source under `macos/Sources` is read and the
/// literals that read as a sentence a contributor would be shown are
/// counted, per file.
///
/// It is a ratchet, not a clean bill of health. This shell today authors
/// most of its own wording: `TCShellCore`'s `*Copy` types were transcribed
/// from the same shared design the Windows interop classes were, and the
/// SwiftUI views carry their own labels and help text. Moving all of that
/// behind the ABI is a project of its own -- see the migration spec. Every
/// file that authors wording is recorded below with the exact number it
/// holds today, so nothing new can be added to it and nothing new can start
/// doing it.
final class ShellWordingTests: XCTestCase {
    /// Files that author wording today, and exactly how much.
    ///
    /// TODO(shell-copy): every entry here is a file whose wording should be
    /// composed in the Rust contributor crate and read across the C ABI, the
    /// way `routing_copy.rs` and `witness_copy.rs` already are. Until then
    /// the number is a CEILING AND A FLOOR both: adding a sentence fails,
    /// and removing one fails too, so the entry has to be lowered
    /// deliberately as copy moves out. Never raise a number. A new file must
    /// never be added here.
    ///
    /// MEASURED, NOT ESTIMATED. Every number came from this file's own
    /// scanner via TC_WORDING_DUMP=1; none was typed by hand.
    private static let wordingBaseline: [String: Int] = [
        // The bridge. One error string apiece; nothing a contributor reads.
        "TCBridge/TCDaemon.swift": 2,

        // TCShellCore's copy types -- transcribed from the same shared design
        // the Windows interop classes were, and the first thing Slice 1 moves.
        "TCShellCore/ArmingOffer.swift": 4,
        "TCShellCore/ContributorVerdict.swift": 4,
        "TCShellCore/CorrectionCopy.swift": 4,
        "TCShellCore/DailyBudgetCopy.swift": 6,
        "TCShellCore/DigestCopy.swift": 3,
        "TCShellCore/MenuBarStatus.swift": 4,
        "TCShellCore/OriginalSearchOutcome.swift": 4,
        "TCShellCore/ProjectArmingCopy.swift": 4,
        "TCShellCore/ProjectIgnoreCopy.swift": 10,
        "TCShellCore/ProjectRow.swift": 3,
        "TCShellCore/RedactionLabels.swift": 3,
        "TCShellCore/RedactionMarks.swift": 2,
        "TCShellCore/RedactionSummary.swift": 7,
        "TCShellCore/ScrubDetectors.swift": 2,
        "TCShellCore/SourceCandidate.swift": 5,
        "TCShellCore/StateDirectory.swift": 2,
        "TCShellCore/SubagentCopy.swift": 6,
        "TCShellCore/SubmitToast.swift": 10,

        // The app model and its non-view surfaces.
        "TraceCommonsApp/AppDelegate.swift": 1,
        "TraceCommonsApp/AppModel.swift": 8,
        "TraceCommonsApp/HealthCopy.swift": 32,
        "TraceCommonsApp/Notifier.swift": 3,
        "TraceCommonsApp/SelfTest.swift": 15,

        // The SwiftUI views, which carry their own labels and help text.
        "TraceCommonsApp/Views/ActionErrorBanner.swift": 2,
        "TraceCommonsApp/Views/BrandMark.swift": 1,
        "TraceCommonsApp/Views/ConsentScopesView.swift": 7,
        "TraceCommonsApp/Views/CreditRecordView.swift": 9,
        "TraceCommonsApp/Views/HistoryView.swift": 26,
        "TraceCommonsApp/Views/MainWindowView.swift": 14,
        "TraceCommonsApp/Views/MenuBarView.swift": 11,
        "TraceCommonsApp/Views/OnboardingConnectView.swift": 7,
        "TraceCommonsApp/Views/OnboardingCoordinatorView.swift": 5,
        "TraceCommonsApp/Views/OnboardingDoneView.swift": 8,
        "TraceCommonsApp/Views/OnboardingPrivacyScanView.swift": 5,
        "TraceCommonsApp/Views/OnboardingProjectsView.swift": 4,
        "TraceCommonsApp/Views/OnboardingRootsView.swift": 5,
        "TraceCommonsApp/Views/OnboardingWelcomeView.swift": 8,
        "TraceCommonsApp/Views/PreviewSheet.swift": 38,
        "TraceCommonsApp/Views/PublicProfileCopy.swift": 46,
        "TraceCommonsApp/Views/QueueFolderRow.swift": 3,
        "TraceCommonsApp/Views/QueueView.swift": 26,
        "TraceCommonsApp/Views/ScrubbingCaveat.swift": 4,
        "TraceCommonsApp/Views/SettingsView.swift": 40,
        "TraceCommonsApp/Views/WhatGetsRemovedSheet.swift": 4,
        "TraceCommonsApp/Views/WithdrawalCopy.swift": 55,
    ]

    /// The surfaces whose wording already comes from Rust. Nothing may ever
    /// buy them an allowance in the baseline: they hold no sentence at all,
    /// and an entry here would be a quiet way of undoing that.
    private static let rustOwnedSurfaces = [
        "TCBridge/TCConsentCopy.swift",
        "TCBridge/TCRoutingCopy.swift",
        "TCShellCore/ConsentCopy.swift",
        "TCShellCore/ReadGate.swift",
        "TCShellCore/RoutingCopy.swift",
        "TCShellCore/RoutingSurface.swift",
    ]

    /// Words a sentence has and an identifier, a wire key, a symbol name or
    /// a format pattern does not.
    ///
    /// The same list the Windows guard uses, deliberately: two shells
    /// counting the same corpus by different rules would produce two numbers
    /// nobody could compare. It is a function-word test rather than a "has a
    /// space" test, because `"chevron.right"`, `"MMMM d, yyyy"` and
    /// `"en_US_POSIX"` are not wording and `"Watch this folder"` is.
    private static let functionWords: Set<String> = Set(
        """
        a an and are as at be been being but by can cannot could did do does for from
        had has have how if in into is isn't it it's its just may never no not nothing of off on once only or
        so some still such than that the their them then there they this those to until up was we were what
        when where which while who will with would you your yours yet anything something everything already
        always about after again all any because before both each else ever every here more most much
        must need needs same see should since take takes tell these too under use used using very
        """
        .split(whereSeparator: { $0 == " " || $0 == "\n" })
        .map(String.init))

    /// No file in this shell authors more wording than it did when this
    /// guard was written, and no file starts authoring wording that did not.
    func testNoWordingIsAuthoredInThisShellBeyondTheRecordedBaseline() {
        let scanned = ShellSources.scan()

        // A scan that found nothing would turn this test into a pass over
        // nothing, which is the failure mode the Windows guard names
        // explicitly. There are 96 Swift sources under macos/Sources today.
        XCTAssertGreaterThanOrEqual(
            scanned.count, 86,
            "only \(scanned.count) Swift sources were scanned under \(ShellSources.root().path); "
                + "the whole tree is expected")

        if ProcessInfo.processInfo.environment["TC_WORDING_DUMP"] == "1" {
            for path in scanned.keys.sorted() where !scanned[path]!.isEmpty {
                print("        \"\(path)\": \(scanned[path]!.count),")
            }
        }

        var failures: [String] = []
        for path in scanned.keys.sorted() {
            let wording = scanned[path]!
            let allowed = Self.wordingBaseline[path] ?? 0
            if wording.count == allowed { continue }
            if wording.count > allowed {
                failures.append(
                    "\(path): \(wording.count) authored sentences, baseline allows \(allowed). "
                        + "First one over the line: \"\(wording[allowed])\"")
            } else {
                failures.append(
                    "\(path): \(wording.count) authored sentences, baseline still allows "
                        + "\(allowed). Wording moved out -- lower the entry (or delete it at zero).")
            }
        }
        for recorded in Self.wordingBaseline.keys.sorted() where scanned[recorded] == nil {
            failures.append(
                "\(recorded): recorded in the baseline but no longer in the shell. Delete the entry.")
        }

        XCTAssertTrue(
            failures.isEmpty,
            "Wording on this shell's surfaces comes from the Rust contributor crate across the "
                + "ABI.\nA sentence written here is one the other two shells will not get, and one "
                + "a rename in the Rust will not reach.\n\n" + failures.joined(separator: "\n"))
    }

    /// The surfaces the Rust already owns hold no wording at all, and hold
    /// no baseline entry either.
    func testTheRustOwnedSurfacesAreNotGivenAWordingAllowance() {
        let scanned = ShellSources.scan()
        for surface in Self.rustOwnedSurfaces {
            XCTAssertNil(
                Self.wordingBaseline[surface],
                "\(surface) has a wording baseline entry. Its wording comes from Rust; an "
                    + "allowance here would quietly undo that.")
            guard let wording = scanned[surface] else {
                XCTFail("\(surface) was not among the scanned sources; the guard would pass over nothing.")
                continue
            }
            XCTAssertEqual(wording, [], "\(surface) authors wording: \(wording)")
        }
    }
}

/// Every Swift source of this shell, and the sentences each one authors.
private enum ShellSources {
    /// `.../macos/Tests/TCShellCoreTests/ShellWordingTests.swift` ->
    /// `.../macos/Sources`. Located from the source file's own path, the way
    /// `ScrubDetectorsTests` locates the shared scrub-label fixture: the
    /// working directory of `swift test` is not something to depend on.
    static func root() -> URL {
        URL(fileURLWithPath: #filePath)
            .deletingLastPathComponent()  // TCShellCoreTests
            .deletingLastPathComponent()  // Tests
            .deletingLastPathComponent()  // macos
            .appendingPathComponent("Sources")
    }

    /// Keyed by path relative to `macos/Sources`, so an entry reads the way
    /// the Windows baseline's do.
    static func scan() -> [String: [String]] {
        let base = root()
        var scanned: [String: [String]] = [:]
        guard
            let walker = FileManager.default.enumerator(
                at: base, includingPropertiesForKeys: nil)
        else {
            return scanned
        }
        for case let url as URL in walker where url.pathExtension == "swift" {
            let relative = url.path
                .replacingOccurrences(of: base.path + "/", with: "")
            guard let text = try? String(contentsOf: url, encoding: .utf8) else { continue }
            scanned[relative] = authoredWording(in: text)
        }
        return scanned
    }

    /// The sentences a Swift source authors.
    ///
    /// A character walk rather than a regex, because this shell writes copy
    /// in `"""` blocks and a line-oriented scan reads one of those as three
    /// unterminated literals. Comments go first, for the reason the routing
    /// guard gives: prose about the wire may quote it, and nothing in a
    /// comment is rendered.
    ///
    /// Known and accepted limitation: an interpolation is folded away with
    /// its escape, so `"\(n) sessions"` is read as `"n) sessions"`. That is
    /// deterministic, and the baseline is measured with it, so it ratchets
    /// correctly -- it is not a claim that interpolated sentences are not
    /// wording.
    static func authoredWording(in source: String) -> [String] {
        let chars = Array(source)
        var literals: [(text: String, line: String)] = []
        var i = 0
        var lineStart = 0

        func currentLine() -> String {
            var end = lineStart
            while end < chars.count && chars[end] != "\n" { end += 1 }
            return String(chars[lineStart..<end])
        }

        while i < chars.count {
            if chars[i] == "\n" {
                i += 1
                lineStart = i
                continue
            }
            // Line comment.
            if chars[i] == "/" && i + 1 < chars.count && chars[i + 1] == "/" {
                while i < chars.count && chars[i] != "\n" { i += 1 }
                continue
            }
            // Block comment, nested the way Swift allows.
            if chars[i] == "/" && i + 1 < chars.count && chars[i + 1] == "*" {
                var depth = 1
                i += 2
                while i + 1 < chars.count && depth > 0 {
                    if chars[i] == "/" && chars[i + 1] == "*" { depth += 1; i += 2; continue }
                    if chars[i] == "*" && chars[i + 1] == "/" { depth -= 1; i += 2; continue }
                    if chars[i] == "\n" { lineStart = i + 1 }
                    i += 1
                }
                continue
            }
            // Multiline literal.
            if chars[i] == "\"" && i + 2 < chars.count && chars[i + 1] == "\"" && chars[i + 2] == "\"" {
                let line = currentLine()
                i += 3
                var text = ""
                while i + 2 < chars.count
                    && !(chars[i] == "\"" && chars[i + 1] == "\"" && chars[i + 2] == "\"")
                {
                    if chars[i] == "\\" { i += 2; continue }
                    text.append(chars[i] == "\n" ? " " : chars[i])
                    i += 1
                }
                i += 3
                literals.append((text, line))
                continue
            }
            // Single-line literal.
            if chars[i] == "\"" {
                let line = currentLine()
                i += 1
                var text = ""
                while i < chars.count && chars[i] != "\"" {
                    if chars[i] == "\\" { i += 2; continue }
                    text.append(chars[i])
                    i += 1
                }
                i += 1
                literals.append((text, line))
                continue
            }
            i += 1
        }

        // A message handed to fatalError, an assertion or a log is read by
        // whoever is debugging this and by nobody else. Holding those to the
        // shared copy would be a refinement of nothing.
        let developerFacing = [
            "fatalError", "preconditionFailure", "assertionFailure", "precondition(",
            "assert(", "print(", "NSLog", "os_log", "Logger(", "logger.",
        ]
        return literals
            .filter { literal in !developerFacing.contains { literal.line.contains($0) } }
            .map(\.text)
            .filter(readsAsASentence)
    }

    /// True where the literal reads as a sentence somebody wrote for a
    /// contributor to read.
    static func readsAsASentence(_ literal: String) -> Bool {
        let text = literal.replacingOccurrences(of: "\n", with: " ")
        guard text.contains(" ") else { return false }
        let words = text
            .split(separator: " ")
            .map { token in
                String(token.filter { $0.isLetter || $0 == "'" }).lowercased()
            }
            .filter { !$0.isEmpty }
        guard words.count >= 2 else { return false }
        return words.contains { ShellWordingTests.functionWordsContains($0) }
    }
}

extension ShellWordingTests {
    /// Exposed so `ShellSources` can reach the one list both shells share.
    fileprivate static func functionWordsContains(_ word: String) -> Bool {
        functionWords.contains(word)
    }
}
